rfd/0073-discover.md

```diff
@@ -0,0 +1,256 @@
+---
+authors: Roman Tkachenko (roman@goteleport.com)
+state: draft
+---
+
+# RFD 73 - Teleport Discover
+
+## Required approvers
+
+@klizhentas && @xinding33 && @jimbishopp
+
+## What
+
+Proposes the `teleport discover` command that simplifies the UX for the
+first-time users who are connecting their cloud resources to Teleport.
+
+## Related RFDs
+
+- [RFD 38: AWS databases auto-discovery](./0038-database-access-aws-discovery.md)
+- [RFD 57: AWS EC2 nodes auto-discovery](https://github.com/gravitational/teleport/pull/12410)
+
+## Why
+
+The proposal is aimed at providing an easy way for Teleport administrators to
+connect their cloud-hosted resources (EC2 instances, RDS databases, EKS clusters
+an so on) to a Teleport cluster.
+
+Over the past few releases Teleport has been adding automatic discovery
+capabilities allowing it to find and register AWS databases, EC2 instances
+and (WIP as of this writing) EKS clusters. Despite providing an improved UX
+compared to registering the resources manually, connecting resources to cluster
+and setting up auto-discovery remains cumbersome with multiple different
+`teleport configure` commands and does not provide good visibility into the
+discovered resources.
+
+The `teleport discover` approach aims to take advantage of the existing auto
+discovery mechanisms Teleport has and provide a unified approach for users to
+enroll their cloud resources.
+
+## Scope
+
+- Works with both self-hosted edition and Teleport Cloud.
+- Focuses on discovering resources in AWS, other clouds to follow.
+- Covers currently supported cloud resources: EC2 instances and RDS databases.
+
+## Prerequisites
+
+In order to use `teleport discover` the user will need:
+
+- Running Auth and Proxy, or a Cloud account.
+- An AWS EC2 instance to run `teleport discover` command from. The node must have
```

Comment:
Is it possible to drop the requirement of having an EC2 machine to run this?

If possible, this will be a huge improvement for the EKS enrollment process. 

Right now the IAM Role mapping into Kubernetes Role is done by appending extra configuration in the `configmap/aws-auth`. 
If  `teleport discover` is running in an EC2 machine, it requires that the operator must add, manually, the EC2 node role into `configmap/aws-auth`  which does not solve our issues. 

If we drop the requirement for at least the Kubernetes discovery, we can run the EKS discovery using the tokens from the user, we can grab the `AWS_PROFILE`, and it will inherit the user credentials giving Teleport the required access level to the cluster. For this solution, we should install the helm chart in the cluster. 

The `teleport discovery` is not required to start the daemon process, it can run while Teleport installs the Helm charts and exits after every cluster is ready. If the operator wants to enroll more clusters he runs the command multiple times.

Otherwise, if the operator wants to run with `--enable-discovery`, he must manually edit the `configmap/aws-auth` in order to map the EC2 IAM role into the Kubernetes role but only for the new clusters, since the old ones were already enrolled using his credentials.

What do you think?

Reply:
From a UX standpoint, it'd be great if we can remove the EC2 requirement but then where would we expect users to run these commands and what permissions (and accompanying security scope) would we inherit?

Reply:
My initial idea was to use `teleport discover aws` to simplify the current infrastructure onboarding with a minimal amount of actions even for cases where the infrastructure is static and continuous discovery is not something that fits the client.

When running the CLI locally, it  could have two modes:

- Automatic: After the administrator decided which resources to onboard, Teleport inherits the user credentials and does everything automatically. This includes the setup of required IAM policies for Teleport to fully work. Although the process is automatic the actions are shown to the user so he can review them. 
```
[1] Connecting to the AWS environment...
[2] Checking your user permissions....
...
[n] Onboarding database...
The teleport cluster lacks the following permissions to onboard the database ...
`json policy`

Do you accept the automatic creation of the policy?
(Y) Yes
(M) Manual
(N) No

...
```
- Manual:  we show the action required as Sasha mentioned:
```go
$ aws iam assign policy < `json policy`

or use this IAM JSON blob...
```

The same applies to other resources like EKS clusters (edit of `configmap/aws-auth`), nodes... 

Using Automatic mode has several security concerns, but if we make it in a way where users control the process  and they run it locally for short periods of time it would simplify a lot the onboarding process.